actor/todo-sql/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+    let re = regex::Regex::new(r"^[a-zA-Z0-9 ,._-/]+$").unwrap();
+    re
+});
+
+#[derive(Serialize, Deserialize)]
+struct InputTodo {
+    title: String,
+    order: Option<i32>,
+}
+#[derive(Serialize, Deserialize)]
+struct UpdateTodo {
+    title: Option<String>,
+    completed: Option<bool>,
+    order: Option<i32>,
+}
+// because 'order' is a keyword in sql, we've renamed the field 'priority',
+// but for serde serialization, leave it as 'order' so that it works
+// with existing javascript written for this app
+#[derive(Serialize, Deserialize, Decode)]
+struct Todo {
+    #[n(0)] url: String,
+    #[n(1)] title: String,
+    #[n(2)] completed: bool,
+    #[serde(rename="order")]
+    #[n(3)] priority: i32,
+}
+
+impl Todo {
+    fn new(url: String, title: String, priority: i32) -> Self {
+        Self {
+            url,
+            title,
+            completed: false,
+            priority,
+        }
+    }
+
+    fn update(self, update: UpdateTodo) -> Todo {
+        Todo {
+            url: self.url,
+            title: update.title.unwrap_or(self.title),
+            completed: update.completed.unwrap_or(self.completed),
+            priority: update.order.unwrap_or(self.priority),
+        }
+    }
+}
+
+// check safety of any user input string before it used in an sql statement
+// If the provider (&interface) supported prepared statements,
+// this could allow more characters
+fn check_safety(tag: &str, uncertain_input: &str) -> Result<(), SqlDbError> {
+    if !REGEX.is_match(&uncertain_input) {
+        return Err(SqlDbError::new("invalid",
+           format!("{} contains invalid characters", tag)))
+    }
+    Ok(())
+}
+
+/// create an empty table with the proper schema
+async fn create_table(ctx: &Context, client: &Db) -> Result<(),SqlDbError> {
+
+    let sql = format!(r#"create table {} (
+            id varchar(36) not null,
+            url varchar(42) not null,
+            title varchar(100) not null,
+            priority int4 not null default 0,
+            completed bool not null default false
+        );"#, TABLE_NAME);
+    let _resp = client.execute(&ctx, &sql) .await?;
+    Ok(())
+}
+
+/// insert a new to-do item
+async fn create_todo(ctx: &Context, client: &Db, input: InputTodo) -> Result<Todo,SqlDbError> {
+
+    check_safety("title", &input.title)?;
+    let id = generate_guid().await.unwrap();
+    let todo = Todo::new(
+        format!("/api/{}", id),
+        input.title,
+        input.order.unwrap_or(0),
+    );
+
+    let _resp = client
+        .execute(&ctx, &format!(r#"
+        insert into {} (id, url, title, order, completed) values ('{}','{}', '{}', {})
+        "#, TABLE_NAME, id, &todo.url, &todo.title, todo.priority))
+        .await?;
+
+    Ok(todo)
+}
+
+/// updates an existing to-do item
+async fn update_todo(ctx: &Context, client: &Db, url: &str, update: UpdateTodo) -> Result<Todo, SqlDbError> {
+    info!("Updating a todo...");
+
+    check_safety("url", url)?;
```

Comment:
per previous comment - this isn't something (imho) developers should have to do explicitly. The provider should deal with this for them.

Reply:
agreed - this is a work-around to accommodate lack of prepared statements, which will be in a future version.